onnxruntime/core/session/inference_session.cc

```diff
@@ -1578,51 +1590,74 @@ common::Status InferenceSession::Initialize() {
       // now that all the transforms are done, call Resolve on the main graph. this will recurse into the subgraphs.
       ORT_RETURN_IF_ERROR_SESSIONID_(graph.Resolve());
 
-      // Currently only the CUDA EP is considered.
+      // Currently CUDA graph is only considered by CUDA EP and TRT EP.
       // If the CUDA EP is part of the providers list for this session AND
       // The CUDA EP is configured to do a graph capture AND
-      // All the graph nodes have been assigned to the CUDA EP,
+      // All the compute graph nodes have been assigned to the CUDA EP,
       // Then the CUDA EP is cached for triggering a ReplayGraph() in Run().
-      auto* cuda_ep = execution_providers_.Get(onnxruntime::kCudaExecutionProvider);
-      if (cuda_ep && cuda_ep->IsGraphCaptureEnabled()) {
-        if (HasControlflowNodes(graph)) {
-          LOGS(*session_logger_, ERROR) << "This session cannot use the CUDA Graph feature as requested by the user "
-                                        << " as the model has control flow nodes which can't be supported by CUDA Graphs.";
-
-          // Return error status as we don't want the session initialization to complete successfully
-          // if the user has requested usage of CUDA Graph feature and we cannot honor that.
-          ORT_RETURN_IF_ERROR_SESSIONID_(
-              ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
-                              "This session cannot use the CUDA Graph feature as requested by the user "
-                              " as the model has control flow nodes which can't be supported by CUDA Graphs."));
-        }
+      // Similar logic is applied to TRT EP.
+      std::vector<const char*> cuda_graph_support_ep_list = {onnxruntime::kTensorrtExecutionProvider, onnxruntime::kCudaExecutionProvider};
+
+      for (auto& it : cuda_graph_support_ep_list) {
```

Comment:
What is the expected behavior when a user registers both the TRT and the CUDA EPs (and they set graph capture to true for both EPs) ?
If I understand correctly, the code path below will first look to see if the TRT EP can capture the graph - if it can then all is good, but if it can't then it will return an error status back to the user ? Do you want to check if the CUDA EP can capture the graph before returning the error status to the user ?

Reply:
Re:  What is the expected behavior when a user registers both the TRT and the CUDA EPs (and they set graph capture to true for both EPs) ?
That's a good question. My thinking of expected behavior is either (1) all graph nodes assigned to TRT EP and TRT EP will run with cuda graph or (2) only a portion of graph nodes assigned to TRT EP, ORT throws exception to let users know and expect them to disable cuda graph for TRT EP.

Re: Do you want to check if the CUDA EP can capture the graph before returning the error status to the user ?
I think we can still do this, but just wondering whether this will be a real case where users register TRT/CUDA EP with both cuda graph enabled and end up only using CUDA EP with cuda graph. Most users registered TRT EP will want to use TRT EP and also, it's highly possible TRT/CUDA EP without cuda graph has better performance than only CUDA EP with cuda graph feature.

Reply:
Cool, thanks. If it is not a common scenario, then we probably don't need to address it right now.